Kernel/WorkQueue.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2021, The SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/IntrusiveList.h>
+#include <Kernel/Forward.h>
+
+namespace Kernel {
+
+extern WorkQueue* g_io_work;
+
+class WorkQueue {
+    AK_MAKE_NONCOPYABLE(WorkQueue);
+    AK_MAKE_NONMOVABLE(WorkQueue);
+
+public:
+    static void initialize();
+
+    WorkQueue(const char*);
+
+    void queue(void (*function)(void*), void* data = nullptr, void (*free_data)(void*) = nullptr)
```

Comment:
Instead of using a C API it seems more "serenity like" to queue an object (Maybe called `WorkItem`?) that has an virtual `execute()`/`do_work()` method on it instead? Then callers can just derive their workitem types in a normal C++ fashion?

Reply:
Since work queues are generally used to execute arbitrary code queuing a function or functor (see the overload below this) makes more sense to me. This way you don't have to come up with objects just to be able to execute some code. This allows you to queue lambdas very easily.

Reply:
I just made a commit with fixes to `WorkQueue` code. Hopefully it will be enough for merging, but suggestions are always appreciated :)

Reply:
FWIW I prefer the implementation I had before. I purposely added the c-style API for the free function/lambda case that doesn't require data, similar to the deferred call api.

Reply:
Let's revert this back then to what it was. If we need to, we can always come up with a better idea what to do :)